### PR TITLE
Ignores a default subclass when `becomes(Parent)`

### PR DESCRIPTION
**Pull request by leonelgalan**
### Summary

Fixes issue described in #30399: A default value on the
inheritance column prevented `child.becomes(Parent)` to return
an instance of `Parent` as expected, instead it returns an instance
of the default subclass.

The change was introduced by #17169 and it was meant to affect
initialization, alone. Where `Parent.new` is expected to return
an instance of the default subclass.

### CODE REVIEW

**Comment by leonelgalan** [comment]
After the original issue was marked as stale I've confirmed this is still an issue in Rails 5.1.4 and 5.2.0.beta2. I rebased this PR to make it easier for reviewers.

File: activerecord/lib/active_record/persistence.rb [hunk 1]
```diff
@@ -359,7 +359,12 @@ def destroy!
     # Any change to the attributes on either instance will affect both instances.
     # If you want to change the sti column as well, use #becomes! instead.
     def becomes(klass)
-      became = klass.new
+      if klass.descends_from_active_record? &&
+         klass.attribute_names.include?(klass.inheritance_column)
+        became = klass.new(klass.inheritance_column => klass.name)
+      else
+        became = klass.new
+      end
```
Comment (kamipo) [inline activerecord/lib/active_record/persistence.rb]: How about avoiding `klass.new` simply?

```diff
diff --git a/activerecord/lib/active_record/persistence.rb b/activerecord/lib/active_record/persistence.rb
index 462e5e7aaf..e5fa4a925e 100644
--- a/activerecord/lib/active_record/persistence.rb
+++ b/activerecord/lib/active_record/persistence.rb
@@ -359,7 +359,8 @@ def destroy!
     # Any change to the attributes on either instance will affect both instances.
     # If you want to change the sti column as well, use #becomes! instead.
     def becomes(klass)
-      became = klass.new
+      became = klass.allocate
+      became.send(:initialize)
       became.instance_variable_set("@attributes", @attributes)
       became.instance_variable_set("@mutations_from_database", @mutations_from_database) if defined?(@mutations_from_database)
       became.instance_variable_set("@changed_attributes", attributes_changed_by_setter)
```

Comment (leonelgalan) [reply activerecord/lib/active_record/persistence.rb]: I'm not familiar with the cons of `klass.new` and `.allocate`. Let me read more about it and I'll change the code.

**Review by kamipo — APPROVED** [review]
Looks good to me. Can you squash your commits into one?

**Comment by leonelgalan** [comment]
Squashed and rebased, thanks for your help!

**Pull request merged**